pkg/compactor/visit_marker.go

```diff
@@ -0,0 +1,194 @@
+
+	"github.com/cortexproject/cortex/pkg/util/runutil"
+)
+
+var (
+	errorVisitMarkerNotFound  = errors.New("visit marker not found")
+	errorUnmarshalVisitMarker = errors.New("unmarshal visit marker JSON")
+)
+
+type VisitStatus string
+
+const (
+	Pending    VisitStatus = "pending"
+	InProgress VisitStatus = "inProgress"
+	Completed  VisitStatus = "completed"
+	Failed     VisitStatus = "failed"
+)
+
+type VisitMarker interface {
+	GetVisitMarkerFilePath() string
+	UpdateStatus(ownerIdentifier string, status VisitStatus)
+	GetStatus() VisitStatus
+	IsExpired(visitMarkerTimeout time.Duration) bool
+	String() string
+}
+
+type VisitMarkerManager struct {
+	bkt                    objstore.InstrumentedBucket
+	logger                 log.Logger
+	ownerIdentifier        string
+	visitMarker            VisitMarker
+	visitMarkerReadFailed  prometheus.Counter
+	visitMarkerWriteFailed prometheus.Counter
+}
+
+func NewVisitMarkerManager(
+	bkt objstore.InstrumentedBucket,
+	logger log.Logger,
+	ownerIdentifier string,
+	visitMarker VisitMarker,
+	visitMarkerReadFailed prometheus.Counter,
+	visitMarkerWriteFailed prometheus.Counter,
+) *VisitMarkerManager {
+	return &VisitMarkerManager{
+		bkt:                    bkt,
+		logger:                 log.With(logger, "type", fmt.Sprintf("%T", visitMarker)),
+		ownerIdentifier:        ownerIdentifier,
+		visitMarker:            visitMarker,
+		visitMarkerReadFailed:  visitMarkerReadFailed,
+		visitMarkerWriteFailed: visitMarkerWriteFailed,
+	}
+}
+
+func (v *VisitMarkerManager) HeartBeat(ctx context.Context, errChan <-chan error, visitMarkerFileUpdateInterval time.Duration, deleteOnExit bool) {
+	level.Info(v.getLogger()).Log("msg", "start visit marker heart beat")
+	ticker := time.NewTicker(visitMarkerFileUpdateInterval)
+	defer ticker.Stop()
+heartBeat:
+	for {
+		v.MarkInProgress(ctx)
+
+		select {
+		case <-ctx.Done():
+			level.Warn(v.getLogger()).Log("msg", "visit marker heart beat got cancelled")
+			v.MarkPending(context.Background())
+			break heartBeat
+		case <-ticker.C:
+			continue
+		case err := <-errChan:
+			if err == nil {
+				level.Info(v.getLogger()).Log("msg", "update visit marker to completed status")
+				v.MarkCompleted(ctx)
+			} else {
+				level.Warn(v.getLogger()).Log("msg", "stop visit marker heart beat due to error", "err", err)
+				if compact.IsHaltError(err) {
+					level.Info(v.getLogger()).Log("msg", "update visit marker to failed status", "err", err)
+					v.MarkFailed(ctx)
+				} else {
+					level.Info(v.getLogger()).Log("msg", "update visit marker to pending status", "err", err)
+					v.MarkPending(ctx)
+				}
+			}
+			break heartBeat
+		}
+	}
+	level.Info(v.getLogger()).Log("msg", "stop visit marker heart beat")
+	if deleteOnExit {
+		level.Info(v.getLogger()).Log("msg", "delete visit marker when exiting heart beat")
+		v.DeleteVisitMarker(context.Background())
+	}
+}
+
+func (v *VisitMarkerManager) MarkInProgress(ctx context.Context) {
```

Comment:
Should we consolidate `MarkXXX` functions into one function with support of different status?